From: Horst Hummel <horst.hummel@de.ibm.com>
Subject: fdasd failing on some images when installing sles10
References: 146313 - LTC21059

workaround - until final fix is available

Since this problem is blocking lots of other tests and I did not find a final
solution so far, I will post a patch that retries to get a correct bdev struct.

This works on my machine and on our tester\'s machine, BUT it is no final fix!!

As soon as a final solution is found, this workaround _must_ be replaced by the
final patch!! 

Acked-by: Hannes Reinecke <hare@suse.de>

--- linux-2.6.16/drivers/s390/block/dasd_genhd.c.orig	2006-02-24 12:59:29.000000000 +0100
+++ linux-2.6.16/drivers/s390/block/dasd_genhd.c	2006-02-24 14:08:37.000000000 +0100
@@ -17,6 +17,7 @@
 #include <linux/fs.h>
 #include <linux/blkpg.h>
 
+#include <asm/delay.h>
 #include <asm/uaccess.h>
 
 /* This is ugly... */
@@ -100,12 +101,21 @@ int
 dasd_scan_partitions(struct dasd_device * device)
 {
 	struct block_device *bdev;
+	int i;
 
 	/* Make the disk known. */
 	set_capacity(device->gdp, device->blocks << device->s2b_shift);
-	bdev = bdget_disk(device->gdp, 0);
-	if (!bdev || blkdev_get(bdev, FMODE_READ, 1) < 0)
-		return -ENODEV;
+
+	for (i=0; i < 512; i++) {
+		bdev = bdget_disk(device->gdp, 0);
+		if (!bdev || blkdev_get(bdev, FMODE_READ, 1) < 0)
+			return -ENODEV;
+		if(bdev->bd_inode->i_size)
+			break;
+		blkdev_put(bdev);
+		udelay(1000);
+	}
+	
 	/*
 	 * See fs/partition/check.c:register_disk,rescan_partitions
 	 * Can't call rescan_partitions directly. Use ioctl.
